### PR TITLE
Adds login function

### PR DESCRIPTION
**Pull request by kimong1**
I just added `logIn` function property to `users`
It includes valdation / encryption then returns user data.
Tests are ready.
Please review.

### CODE REVIEW

File: lib/users.js [hunk 1]
```diff
@@ -8,12 +11,45 @@ const schema = yup.object().shape({
...
+users.signUp = async (userInfo) => {
   return schema.validate(userInfo).catch((err) => { return {
       errors: err.errors
     }
   })
 } 
 
+users.logIn = async (userInfo) => {
+  try {
+    await schema.validate(userInfo)
```
Comment (songz) [inline lib/users.js:23]: remove. User input may be a username or email.

Comment (songz) [reply lib/users.js:23]: you can just do a simple check:

```
if (!userInfo.username && !userInfo.email) {
  logger.info('invalid userInfo for login', userInfo)
  return {
    errors: ['invalid input']
  }
```

File: lib/users.js [hunk 2]
```diff
@@ -8,12 +11,45 @@ const schema = yup.object().shape({
...
+  }
+  catch (err) {
+    logger.info(err)
+    return {
+      errors: err.errors
+    }
+  }
+
+  db.start()
```
Comment (songz) [inline lib/users.js:32]: remove `db.start()`

File: lib/users.test.js [hunk 3]
```diff
@@ -50,3 +50,79 @@ describe('Validates user signup', () => {
     })
   })
 })
+
+jest.mock('../sequelize/db')
+const bcrypt = require('bcrypt')
+const db = require('../sequelize/db')
```
Comment (songz) [inline lib/users.test.js:56]: move these to the top of the file. Other user functions may need to use these. All mocks and requires must be at the top of the file.

File: lib/users.test.js [hunk 4]
```diff
@@ -50,3 +50,79 @@ describe('Validates user signup', () => {
...
+
+const genData = async (password) => {
+  return {
+    username: 'testuser',
+    password: await bcrypt.hash(password, 10),
+    email: 'test.user@databases.com'
+  }
+}
+
+db.start = () => {}
```
Comment (songz) [inline lib/users.test.js:66]: remove

File: lib/users.test.js [hunk 5]
```diff
@@ -50,3 +50,79 @@ describe('Validates user signup', () => {
...
+const mockFindOne = jest.fn()
+db.getModels = () => {
```
Comment (songz) [inline lib/users.test.js:68]: move the db mocks to the top of the file.

File: lib/users.test.js [hunk 6]
```diff
@@ -50,3 +50,79 @@ describe('Validates user signup', () => {
...
+    }
+}
+
+const logIn = users.logIn
+
+beforeEach(() => {
+    jest.clearAllMocks()
+})
+
+describe('Validates user logIn', () => {
```
Comment (songz) [inline lib/users.test.js:82]: This doesn't deserve to be its on suite (every `describe` is a test suite)

File: lib/users.test.js [hunk 7]
```diff
@@ -50,3 +50,79 @@ describe('Validates user signup', () => {
...
+  it("function should reject if username doesn't exist", () => {
```
Comment (songz) [inline lib/users.test.js:83]: `should reject if user input is invalid`

File: lib/users.test.js [hunk 8]
```diff
@@ -50,3 +50,79 @@ describe('Validates user signup', () => {
     })
   })
 })
+
+jest.mock('../sequelize/db')
+const bcrypt = require('bcrypt')
+const db = require('../sequelize/db')
+
+const genData = async (password) => {
```
Comment (greenc123) [inline lib/users.test.js:58]: Instead of genData could you rename it to genTestUserDataWithPassword

Comment (greenc123) [reply lib/users.test.js:58]: genData is kind of confusing

File: lib/users.js (hunk 1 above)
Comment (kimong1) [reply lib/users.js:23]: That's a good idea

File: lib/users.test.js [hunk 9]
```diff
@@ -1,6 +1,31 @@
...
+        Accounts: {
+            findOne: mockFindOne
+        }
+    }
+}
 
 const signUp = users.signUp
+const logIn = users.logIn
+
+beforeEach(() => {
```
Comment (songz) [inline lib/users.test.js:26]: this belongs inside each test suite (aka `describe`)

**Pull request merged**